### PR TITLE
[CPU][DataTiling] Remove EncodingUser attribute, extend encodings to `linalg.generic` ops, remove LiftGenericToTransposeBatchMatmulPass

### PR DESCRIPTION
**Pull request by Max191**
This PR extends data tiling to `linalg.generic` ops that are contractions. This allows materializeEncoding to generate the correct pack ops for any form of generalized batch_matmul operation, avoiding generating extra transpose ops to be handled by Flow. 

Along these lines, the `LiftGenericToTransposeBatchMatmulPass` is removed to prevent creating such transpose ops. In order to handle unsigned semantics for some mmt4d ukernels, the logic for creating a separate producer op to `arith.extui` matmul inputs has been moved from this pass directly to materializeEncoding.

The EncodingUser attribute is also removed, since it is no longer needed after the addition of `user_indexing_maps`.

### CODE REVIEW

**Comment by Max191** [comment]
I think I still need to add some lit tests for generic op encodings.

**Comment by Max191** [comment]
After looking into the failure and looking at `linalg::isaContractionOpInterface` more, I think this needs some stricter checks in setEncoding. `linalg::isaContractionOpInterface` seems to have changed a bit to encompass more types of contractions (like complex) since I last looked at it, and I'm not sure mmt4d can handle those cases. Also there are some other linalg ops that we shouldn't setEncoding on (like `linalg.dot` from the failure). I think it should be good enough to just check the body and have a stricter matching on the sizes of the `indexing_maps`.

File: compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPackUnPack.cpp [hunk 1]
```diff
@@ -107,66 +115,98 @@ RankedTensorType getExpandedType(RankedTensorType type, bool isBatched,
...
+//   auto cDims = linalg::inferContractionDims(linalgOp);
+//   if (cDims->m.size() && cDims->n.size()) {
+//     return value;
+//   }
+//   RankedTensorType type = value.getType().cast<RankedTensorType>();
+//   RankedTensorType newType = getExpandedType(type,
+//   /*isBatched=*/!cDims->batch.empty(), isTransposed, ri); return rewriter
+//       .create<tensor::ExpandShapeOp>(linalgOp->getLoc(), newType, value, ri)
+//       .getResult();
+// }
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoPackUnPack.cpp]: nit: Leftover commented code?

**Comment by hanhanW** [comment]
This is a very big PR, and it is making review harder. Are we able to break it down? Based on the PR title, I feel that it can be 3 independent PRs?

- Remove EncodingUser attribute
- extend encodings to linalg.generic ops
- remove LiftGenericToTransposeBatchMatmulPass

**Comment by Max191** [comment]
This PR is split into https://github.com/openxla/iree/pull/16161, https://github.com/openxla/iree/pull/16163, and https://github.com/openxla/iree/pull/16164

**Pull request closed without merging**